src/dmd/traits.d

```diff
@@ -1594,8 +1594,8 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
 
             if (!err)
             {
-                auto ex = t ? t.typeToExpression() : isExpression(o);
-                if (!ex && t)
+                auto ex = isExpression(o);
+                if (t)
```

Comment:
`resolve` overrides the result of `isExpression` with `null` or another expression

Reply:
Additionally this somehow breaks UFCS in speculative compilations:
```
void visit(alias what, T)(T t) {}

void main()
{
    int number;
    number.visit!(x => x + 1);
    static assert(__traits(compiles, visit!(x => x + 1)(number)));
    static assert(__traits(compiles, number.visit!(x => x + 1))); // fails
}
```

Reply:
Yep, it seems that because the ambiguity in the context of trait-compiles and mixin, the code is parsed as a TypeIdentifier (which doesn't have UFCS) instead of DotIdExp, that's why the type is wrapped with typeToExpression. I've added test code in TypeIdentifier semantics to make it UFCS compatible as an experiment. But overall I'm discarding the approach of using this difference to fix the bug and I'm going to pull a new code.